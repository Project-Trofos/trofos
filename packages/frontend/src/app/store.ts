import { configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';
import trofosApiSlice, { nusmodsApiSlice } from '../api';

const store = configureStore({
  reducer: {
    [trofosApiSlice.reducerPath]: trofosApiSlice.reducer,
    [nusmodsApiSlice.reducerPath]: nusmodsApiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(trofosApiSlice.middleware).concat(nusmodsApiSlice.middleware),
});

// optional, but required for refetchOnFocus/refetchOnReconnect behaviors
// see `setupListeners` docs - takes an optional callback as the 2nd arg for customization
setupListeners(store.dispatch);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
