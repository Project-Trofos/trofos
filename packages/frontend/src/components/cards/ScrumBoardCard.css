.scrum-board-card-container {
  width: 350px;
}

.scrum-board-card-container .ant-card-body {
  padding: 12px;
}

.scrum-board-card-container .backlog-card-id {
  margin-top: 6px;
  width: 52px;
}

.scrum-board-card-container .backlog-card-type {
  width: 50px;
}

.scrum-board-card-container .backlog-card-priority {
  width: 89px;
}

.scrum-board-card-container .backlog-card-assignee {
  width: 69px;
}

.scrum-board-card-container .ant-avatar.ant-avatar-icon {
  font-size: 12px;
}

.scrum-board-card-container textarea.backlog-card-summary.ant-input {
  margin: 0;
}

.scrum-board-card-container .backlog-card-points.ant-input-number {
  width: 46px;
}

.scrum-board-card-details {
  display: flex;
  margin-top: 10px;
}

.scrum-board-card-details span,
.scrum-board-card-details div,
.scrum-board-card-details input,
.scrum-board-card-details .ant-avatar.ant-avatar-icon {
  font-size: 12px;
}

.scrum-board-card-details div {
  margin-right: 3px;
}
