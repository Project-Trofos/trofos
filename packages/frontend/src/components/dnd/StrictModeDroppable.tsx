// Workaround for issue with React 18 strict mode (https://github.com/atlassian/react-beautiful-dnd/issues/2350)
// Code for fix obtained from https://github.com/atlassian/react-beautiful-dnd/issues/2399#issuecomment-1175638194

import React, { useEffect, useState } from 'react';
import { Droppable, DroppableProps } from 'react-beautiful-dnd';

export default function StrictModeDroppable({ children, ...props }: DroppableProps) {
  const [enabled, setEnabled] = useState(false);

  useEffect(() => {
    const animation = requestAnimationFrame(() => setEnabled(true));

    return () => {
      cancelAnimationFrame(animation);
      setEnabled(false);
    };
  }, []);

  if (!enabled) {
    return null;
  }

  return <Droppable {...props}>{children}</Droppable>;
}
