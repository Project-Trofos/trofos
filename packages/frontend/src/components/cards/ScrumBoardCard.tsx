import React from 'react';
import { Card } from 'antd';
import { useNavigate, useParams } from 'react-router-dom';
import { isEqual } from 'lodash';
import { UserAvatar } from '../avatar/UserAvatar';
import type { Backlog } from '../../api/types';
import { draggableWrapper } from '../../helpers/dragAndDrop';
import './ScrumBoardCard.css';

type ScrumBoardCardProps = { backlog: Backlog; projectKey: string | null | undefined };

function Component({ backlog, projectKey }: ScrumBoardCardProps) {
  const navigate = useNavigate();
  const params = useParams();

  return (
    <Card
      className="scrum-board-card-container"
      onClick={() => navigate(`/project/${params.projectId}/backlog/${backlog.backlog_id}`)}
    >
      <div className="scrum-board-card-summary">{backlog.summary}</div>
      <div className="scrum-board-card-details">
        <div className="backlog-card-id">
          {projectKey ? `${projectKey}-` : ''}
          {backlog.backlog_id}
        </div>
        <div className="scrum-board-card-type">{backlog.type}</div>
        <div className={`scrum-board-card-points ${backlog.points ? 'points-active' : ''}`}>{backlog.points}</div>
        <div className={`scrum-board-card-priority ${backlog.priority}-priority`}>{backlog.priority}</div>
        <div className="scrum-board-card-assignee">
          {backlog.assignee && (
            <UserAvatar
              tooltip
              className="assignee-avatar"
              size="small"
              userHashString={backlog.assignee.user.user_email}
              userDisplayName={backlog.assignee.user.user_display_name}
            />
          )}
        </div>
      </div>
    </Card>
  );
}

const ScrumBoardCard = React.memo(draggableWrapper(Component), (prevProps, currProps) => isEqual(prevProps, currProps));
const ReadOnlyScrumBoardCard = React.memo(Component, (prevProps, currProps) => isEqual(prevProps, currProps));

export { ScrumBoardCard, ReadOnlyScrumBoardCard };
