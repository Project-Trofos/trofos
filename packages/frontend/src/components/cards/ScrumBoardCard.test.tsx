import React from 'react';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { DragDropContext } from 'react-beautiful-dnd';
import store from '../../app/store';
import type { Backlog } from '../../api/types';
import ScrumBoardCard from './ScrumBoardCard';
import StrictModeDroppable from '../dnd/StrictModeDroppable';

describe('ScrumBoardCard test', () => {
  const mockBacklog: Backlog = {
    backlog_id: 111,
    summary: 'A Test Summary',
    type: 'story',
    priority: 'very_high',
    sprint_id: 123,
    reporter_id: 1,
    assignee_id: 2,
    points: 3,
    description: 'A test description here',
    project_id: 123,
    status: 'todo',
    assignee: {
      created_at: '2022-12-26T04:19:07.531Z',
      project_id: 903,
      user_id: 2,
      user: {
        user_email: 'testBacklogUser2@test.com',
        user_display_name: 'Backlog User 2',
      },
    },
  };

  const mockProjectKey = 'TEST';
  const mockIndex = 0;
  const mockDroppableId = '0';
  const { baseElement } = render(
    <Provider store={store}>
      <BrowserRouter>
        <DragDropContext onDragEnd={jest.fn()}>
          <StrictModeDroppable droppableId={mockDroppableId}>
            {(provided) => (
              <div ref={provided.innerRef} {...provided.droppableProps}>
                <ScrumBoardCard backlog={mockBacklog} projectKey={mockProjectKey} index={mockIndex} />
              </div>
            )}
          </StrictModeDroppable>
        </DragDropContext>
      </BrowserRouter>
    </Provider>,
  );

  it('renders scrum board card with correct details', () => {
    expect(screen.getByText(mockBacklog.summary)).toBeInTheDocument();
    expect(screen.getByText(mockBacklog.type)).toBeInTheDocument();
    expect(screen.getByText(mockBacklog?.points?.toString() as string)).toBeInTheDocument();

    // Compare with snapshot to ensure structure remains the same
    expect(baseElement).toMatchSnapshot();
  });
});
