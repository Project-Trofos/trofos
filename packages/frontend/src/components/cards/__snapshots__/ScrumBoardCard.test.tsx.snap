// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ScrumBoardCard test > renders scrum board card with correct details 1`] = `
<body>
  <div>
    <div
      data-rbd-droppable-context-id="0"
      data-rbd-droppable-id="0"
    >
      <div
        aria-describedby="rbd-hidden-text-0-hidden-text-0"
        data-rbd-drag-handle-context-id="0"
        data-rbd-drag-handle-draggable-id="111"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="111"
        draggable="false"
        role="button"
        tabindex="0"
      >
        <div
          class="ant-card ant-card-bordered scrum-board-card-container"
        >
          <div
            class="ant-card-body"
          >
            <div
              class="scrum-board-card-summary"
            >
              A Test Summary
            </div>
            <div
              class="scrum-board-card-details"
            >
              <div
                class="backlog-card-id"
              >
                TEST-
                111
              </div>
              <div
                class="scrum-board-card-type"
              >
                story
              </div>
              <div
                class="scrum-board-card-points points-active"
              >
                3
              </div>
              <div
                class="scrum-board-card-priority very_high-priority"
              >
                very_high
              </div>
              <div
                class="scrum-board-card-assignee"
              >
                <span
                  class="ant-avatar ant-avatar-sm ant-avatar-circle assignee-avatar"
                  style="background-color: rgb(105, 193, 165); color: white;"
                >
                  <span
                    class="ant-avatar-string"
                    style="transform: scale(1) translateX(-50%);"
                  >
                    B2
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-atomic="true"
    aria-live="assertive"
    id="rbd-announcement-0"
    style="position: absolute; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%);"
  />
  <div
    id="rbd-hidden-text-0-hidden-text-0"
    style="display: none;"
  >
    
  Press space bar to start a drag.
  When dragging you can use the arrow keys to move the item around and escape to cancel.
  Some screen readers may require you to be in focus mode or to use your pass through key

  </div>
</body>
`;
